pkg/profiler/profile_writer.go

```diff
@@ -59,29 +60,35 @@ func (fw *FileProfileWriter) Write(_ context.Context, labels model.LabelSet, pro
 // RemoteProfileWriter is a profile writer that writes profiles to a remote profile store.
 type RemoteProfileWriter struct {
 	profileStoreClient profilestorepb.ProfileStoreServiceClient
+	// pool of gzip encoders helps to reduce GC pressure.
+	pool sync.Pool
 }
 
 // NewRemoteProfileWriter creates a new RemoteProfileWriter.
 func NewRemoteProfileWriter(profileStoreClient profilestorepb.ProfileStoreServiceClient) *RemoteProfileWriter {
 	return &RemoteProfileWriter{
 		profileStoreClient: profileStoreClient,
+		pool: sync.Pool{New: func() interface{} {
+			z, _ := gzip.NewWriterLevel(nil, gzip.BestSpeed)
+			return z
+		}},
 	}
 }
 
 // Write sends the profile using the designated write client.
 func (rw *RemoteProfileWriter) Write(ctx context.Context, labels model.LabelSet, prof *profile.Profile) error {
 	buf := bytes.NewBuffer(nil)
-	zw, err := gzip.NewWriterLevel(buf, gzip.StatelessCompression)
-	if err != nil {
-		return err
-	}
-	if err = prof.WriteUncompressed(zw); err != nil {
+	zw, _ := rw.pool.Get().(*gzip.Writer)
```

Comment:
ditto

Reply:
Since the pool contains only `*gzip.Writer`, then the type casting should always succeed unless some other object was added to the pool. In that case the program should panic, therefore there mustn't be an underscore. Does it sound reasonable?

Reply:
For sure. I would rather program against the interface and in this case, the contract (interface) tells me that there could be an error. Underlying implementation can change. It's a low-hoop to jump, either propagate or handle the error here. What do you think?